### PR TITLE
Simplifying DafnyServer Code

### PR DESCRIPTION
**Pull request by Dargones**
Some of the code in DafnyServer is either dead or is tailored for processing arbitrary Z3 models generated by Boogie. I propose to simplify this code and make it Dafny-specific by dropping the support for [VCC](https://github.com/microsoft/vcc) and [BCT](http://cs.ioc.ee/etaps12/localproc/bytecode/paper2.pdf). These two projects are, to my understanding, dead and, in any case, don’t have direct ties to Dafny. Making this change would make it much easier to improve counterexample generation and augment it with new features in the future.

None of the changes I propose should modify or augment Dafny's behavior in any way. Here is a detailed description:

- Delete [BaseProvider.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/boogie/BaseProvider.cs) and [VccProvider.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/boogie/VccProvider.cs) which are tailored to support BCT and VCC, respectively.
- Delete dead code (identified by Rider) including [MapletNode](https://github.com/dafny-lang/dafny/blob/5a8e1805083a2a2d3d08365a2cc414d612583843/Source/DafnyServer/boogie/DafnyProvider.cs#L411-L417), [ContainerNode](https://github.com/dafny-lang/dafny/blob/5a8e1805083a2a2d3d08365a2cc414d612583843/Source/DafnyServer/boogie/DataModel.cs#L220-L245) and [TopState](https://github.com/dafny-lang/dafny/blob/5a8e1805083a2a2d3d08365a2cc414d612583843/Source/DafnyServer/boogie/DataModel.cs#L94-L121) classes as well as numerous unused methods and commented-out code in [Namer.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/boogie/Namer.cs), [DataModel.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/boogie/DataModel.cs) and [DafnyProvider.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/boogie/DafnyProvider.cs) files.
- Delete the `GenerateSourceLocations` method from Namer.cs. This method's only side effect is the update of the `sourceLocations` dictionary (also deleted), which is never used elsewhere in the code.
- Delete interfaces in [DataModel.cs](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/boogie/DataModel.cs) since the abstraction they provide is no longer necessary now that the code is Dafny-specific.
- Merge [NamedState](https://github.com/dafny-lang/dafny/blob/8b26545c041cdfa7bdf39381277e1b11a745903c/Source/DafnyServer/boogie/Namer.cs#L487-L523) abstract class with [StateNode](https://github.com/dafny-lang/dafny/blob/8b26545c041cdfa7bdf39381277e1b11a745903c/Source/DafnyServer/boogie/DafnyProvider.cs#L306-L374), which is by this point its only subclass. Call the new class DafnyModelState.
- Use target-typed new where possible and convert get properties to expression-bodied properties for clarity.

The next step would be to merge VariableNode, FieldNode, ElementNode, and DisplayNode classes, which largely duplicate the same functionality, but I think it would be best if I make that and any further changes into separate pull requests.

Thank you!

### CODE REVIEW

File: Source/DafnyServer/boogie/DafnyProvider.cs [hunk 1]
```diff
@@ -298,25 +267,23 @@ Model.Element Unbox(Model.Element elt)
...
+  public class DafnyModelState
   {
     internal readonly DafnyModel dm;
-    public readonly List<VariableNode> Vars = new List<VariableNode>();
-    internal readonly List<VariableNode> skolems;
-    internal readonly int index;
+    public readonly List<VariableNode> Vars = new();
+    private readonly List<VariableNode> skolems;
+    private Model.CapturedState state;
+    private LanguageModel langModel; // no point making it protected - they will need VccModel, DafnyModel
```
Comment (keyboardDrummer) [inline Source/DafnyServer/boogie/DafnyProvider.cs:280]: I find the comment confusing.

**Review by keyboardDrummer — APPROVED** [review]
LGTM

File: Source/DafnyServer/boogie/DafnyProvider.cs (hunk 1 above)
Comment (Dargones) [reply Source/DafnyServer/boogie/DafnyProvider.cs]: Thank you @keyboardDrummer ! I've removed it. The comment wasn't mine - accidentally copied it when moving this line from [here](https://github.com/dafny-lang/dafny/blob/8b26545c041cdfa7bdf39381277e1b11a745903c/Source/DafnyServer/boogie/Namer.cs#L490).

**Review by keyboardDrummer — APPROVED** [review]

File: Source/DafnyServer/boogie/DafnyProvider.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyServer/boogie/DafnyProvider.cs]: I figured as much ^^

**Comment by Dargones** [comment]
Thank you @keyboardDrummer ! Let me know if there is anything else I should do before merging.

**Comment by keyboardDrummer** [comment]
> Thank you @keyboardDrummer ! Let me know if there is anything else I should do before you can approve the merging.

I approved the changes right? Were you able to merge? I'll do it now.

**Comment by Dargones** [comment]
Thank you! I don't have permissions to merge, so I couldn't do that myself .

**Pull request merged**